include/lbann/comm.hpp

```diff
@@ -864,24 +923,24 @@ class lbann_comm {
    */
   int threads_per_proc;
   /** Pre-allocated buffers for collectives. */
-  std::unordered_map<size_t, std::vector<uint8_t *>> collective_bufs;
+  std::map<size_t, std::vector<uint8_t *>> collective_bufs;
   /** Current default allreduce algorithm. */
   allreduce_algorithm default_allreduce_algo =
     allreduce_algorithm::DYNAMIC;
 
 #ifdef LBANN_HAS_ALUMINUM
-  /**
-   * Used to convert between Elemental Comm objects and an Aluminum
-   * communicator.
-   */
-  std::unordered_map<MPI_Comm, allreduces::MPICommunicator*> al_comms;
-
-  /**
-   * Get the Aluminum communicator associated with Elemental communicator c.
-   * This will create a new Al communicator if needed.
-   * @todo This currently only supports the Aluminum MPI backend.
+  using al_comms_key_type = std::pair<MPI_Comm, std::type_index>;
+  using al_comms_val_type = std::unique_ptr<allreduces::MPICommunicator>;
+  std::map<al_comms_key_type, al_comms_val_type> m_al_comms;
```

Comment:
Changed to a `map` since it is non-trivial to make an `unordered_map` use `pair` as keys.